src/metrics.cpp

```diff
@@ -169,17 +170,37 @@ int printStats(bool mining)
     int lines = 4;
 
     int height;
+    int64_t tipmediantime;
     size_t connections;
     int64_t netsolps;
     {
         LOCK2(cs_main, cs_vNodes);
         height = chainActive.Height();
+        tipmediantime = chainActive.Tip()->GetMedianTimePast();
         connections = vNodes.size();
         netsolps = GetNetworkHashPS(120, -1);
     }
     auto localsolps = GetLocalSolPS();
 
-    std::cout << "           " << _("Block height") << " | " << height << std::endl;
+    if (IsInitialBlockDownload()) {
+        // We average the target spacing with the observed spacing to the last
+        // checkpoint, and use that to estimate the current network height.
+        auto chainParams = Params();
+        auto checkpointData = chainParams.Checkpoints();
+        int medianHeight = height - CBlockIndex::nMedianTimeSpan / 2;
+        auto heightLastCheckpoint = Checkpoints::GetTotalBlocksEstimate(checkpointData);
+        auto targetSpacing = chainParams.GetConsensus().nPowTargetSpacing;
+        double checkpointSpacing = (double (tipmediantime - checkpointData.nTimeLastCheckpoint)) / (medianHeight - heightLastCheckpoint);
+        double averageSpacing = (targetSpacing + checkpointSpacing) / 2;
+        int netheight = medianHeight + ((GetTime() - tipmediantime) / averageSpacing);
+        // Round to nearest ten to reduce noise
+        netheight = ((netheight + 5) / 10) * 10;
+        int downloadPercent = height * 100 / netheight;
```

Comment:
Imagine there were some bug in the arithmetic. Could we factor out the "sensor" code (which reads from blockchain state, ex `IsInitialBlockDownload()` and `chainParams.Checkpoints`, arithmetic functional code, and display code, and then unittest at least the arithmetic functional code?

Then if there were an arithmetic bug, that would imply either we missed a test specification, or the specification was buggy.